README.md
# SealDice 插件

## JS扩展指令

### BlackJack 二十一点

经典纸牌游戏. X感骰娘在线发牌.

#### 开发计划

1. 所有人停牌之后自动结算
2. 结算后显示赢家
3. (为了仪式感) 玩家自己抽初始手牌
4. 分牌split

### Inventory 物品栏

记不住背包里有什么? 记不住物品属性/效果/数量? 看这里!

### Timeline 时间线

时间很重要时记不住? 游戏书不知道进行到第几天? 看这里!

## 牌堆

## 自定义回复


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 2023-04-09

### Fixed

1.  修正二十一点结算失败情况下意外造成游戏重置的BUG

## 2023-03-31

### Added

1.  上传之前写的三个扩展指令 `blackjack.js` `inventory.js` `timeline.js`


scripts/blackjack.js
// ==UserScript==
// @name         BlackJack(二十一点)
// @author       JohNSoN
// @version      1.0.1
// @description  经典纸牌游戏Black Jack
// @timestamp    1681024053
// @license      Unlicense
// @homepageURL  https://github.com/Xiangze-Li/sealdice-addon
// ==/UserScript==

const VERSION = '1.0.1';

class Player {
    #name = '';
    #hand = [];
    #stand = false;

    constructor(name, stand, hand = []) {
        this.#name = String(name);
        this.#stand = Boolean(stand);
        if (hand.length > 0) {
            this.#hand = hand;
        }
    }

    toJSON() {
        return {
            name: this.#name,
            deck: this.#hand,
            stand: this.#stand,
        }
    }

    hit(card) {
        if (this.#stand) return false;
        this.#hand.push(card);
        return true;
    }

    stand() {
        this.#stand = true;
    }

    check() {
        if (this.win) {
            if (this.blackjack) return 2;
            return 1;
        }
        if (this.busted) return -1;
        return 0;
    }

    get sum() {
        let s = 0;
        let ace = 0;
        for (let card of this.#hand) {
            let num = card.substring(2);
            switch (num) {
                case 'A':
                    ace++;
                    s += 11;
                    break;
                case 'J': case 'Q': case 'K':
                    s += 10;
                    break;
                default:
                    s += parseInt(num);
                    break;
            }
        }
        while (s > 21 && ace > 0) {
            s -= 10;
            ace--;
        }
        return s;
    }

    get busted() {
        let b = this.sum > 21;
        if (b) this.#stand = true;
        return b;
    }

    get win() {
        let b = this.sum === 21;
        if (b) this.#stand = true;
        return b;
    }

    get blackjack() {
        let b = this.#hand.length === 2 && this.sum === 21;
        if (b) this.#stand = true;
        return b;
    }

    get standed() { return this.#stand; }

    describe(isDealer = false) {
        let s = this.#name + ' ';
        if (isDealer) {
            s += `手牌: [${this.#hand[0]}],[?]`;
            return s;
        }

        s += `手牌: [${this.#hand.join('],[')}]; `;
        s += `点数: ${this.sum}; `;
        switch (this.check()) {
            case 2:
                s += 'BlackJack!';
                break;
            case 1:
                s += '21点!';
                break;
            case -1:
                s += '爆牌!';
                break;
            default:
                s += this.#stand ? '已停牌' : '未停牌';
                break;
        }
        return s;
    }
}

class BlackjackGame {

    #deck = [];
    #players = new Map();
    #dealer = new Player('庄家');
    #status = BlackjackGame.StIdle;

    static OneDeck = [
        '黑桃A', '黑桃2', '黑桃3', '黑桃4', '黑桃5', '黑桃6', '黑桃7', '黑桃8', '黑桃9', '黑桃10', '黑桃J', '黑桃Q', '黑桃K',
        '红桃A', '红桃2', '红桃3', '红桃4', '红桃5', '红桃6', '红桃7', '红桃8', '红桃9', '红桃10', '红桃J', '红桃Q', '红桃K',
        '草花A', '草花2', '草花3', '草花4', '草花5', '草花6', '草花7', '草花8', '草花9', '草花10', '草花J', '草花Q', '草花K',
        '方片A', '方片2', '方片3', '方片4', '方片5', '方片6', '方片7', '方片8', '方片9', '方片10', '方片J', '方片Q', '方片K',
    ];
    static StIdle = 'idle';
    static StShuffled = 'shuffled';
    static StStarted = 'started';
    static StFinished = 'finished';

    addPlayer(id, name = '') {
        if (!name) name = id;
        if (this.#status !== BlackjackGame.StShuffled) return [false, '游戏未准备好'];
        if (this.#players.has(id)) return [false, '玩家已存在'];
        this.#players.set(id, new Player(name, false, []));
        return [true, ''];
    }

    removePlayer(id) {
        if (this.#status !== BlackjackGame.StShuffled) return [false, '游戏未准备好'];
        if (!this.#players.has(id)) return [false, '玩家不存在'];
        this.#players.delete(id);
        return [true, ''];
    }

    start() {
        if (this.#status !== BlackjackGame.StShuffled) return [false, '游戏未准备好'];
        if (this.#players.size <= 0) return [false, '没有玩家'];
        this.#status = BlackjackGame.StStarted;
        this.#dealer.hit(this.#deck.shift());
        this.#dealer.hit(this.#deck.shift());
        for (let pl of this.#players.values()) {
            pl.hit(this.#deck.shift());
            pl.hit(this.#deck.shift());
        }
        return [true, ''];
    }

    hit(plID) {
        if (this.#status !== BlackjackGame.StStarted) return [false, '游戏未开始'];
        if (!this.#players.has(plID)) return [false, '玩家不存在'];
        let pl = this.#players.get(plID);
        if (pl.standed || pl.busted) return [false, '玩家已停牌或爆牌'];
        if (!pl.hit(this.#deck.shift())) return [false, '叫牌出错'];

        return [true, ''];
    }

    stand(plID) {
        if (this.#status !== BlackjackGame.StStarted) return [false, '游戏未开始'];
        if (!this.#players.has(plID)) return [false, '玩家不存在'];
        let pl = this.#players.get(plID);
        if (pl.standed || pl.busted) return [false, '玩家已停牌或爆牌'];
        pl.stand();

        return [true, ''];
    }

    tryConclude() {
        if (this.#status !== BlackjackGame.StStarted) return [false, '游戏未开始'];
        for (let pl of this.#players.values()) {
            if (!pl.standed) return [false, '有玩家未停牌'];
        }

        while (this.#dealer.check() === 0 && this.#dealer.sum < 17) {
            this.#dealer.hit(this.#deck.shift());
        }
        this.#dealer.stand();
        this.#status = BlackjackGame.StFinished;
        return [true, ''];
    }

    status() {
        return this.toJSON();
    }

    describe() {
        let s = '';
        let concluded = false;

        if (this.#status === BlackjackGame.StFinished) {
            s += '游戏已结束\n';
            concluded = true;
        } else if (this.#status === BlackjackGame.StStarted) {
            // s += `剩余牌数：${this.#deck.length}\n`;
        } else {
            s += '游戏未开始';
            return s;
        }
        s += `${this.#dealer.describe(!concluded)}\n`;
        for (let pl of this.#players.values()) {
            s += `${pl.describe()}\n`;
        }
        s.trimEnd();
        return s;
    }

    constructor(str = '') {
        if (!str) {
            this.#deck = BlackjackGame.OneDeck.slice();
            shuffle(this.#deck);
            this.#players = new Map();
            this.#status = BlackjackGame.StShuffled;
            this.#dealer = new Player('骰娘');
            return;
        }

        let obj = JSON.parse(str,
            (key, value) => {
                switch (key) {
                    case 'players':
                        let map = new Map();
                        for (let [k, v] of value) {
                            map.set(k, new Player(v.name, v.stand, v.deck));
                        }
                        return map;
                    case 'dealer':
                        return new Player(value.name, value.stand, value.deck);
                    default:
                        return value;
                }
            }
        );
        this.#deck = obj.deck;
        this.#players = obj.players;
        this.#status = obj.status;
        this.#dealer = obj.dealer;
    }

    toJSON() {
        return {
            deck: this.#deck,
            players: Array.from(this.#players.entries()),
            dealer: this.#dealer,
            status: this.#status,
        }
    }
}


let ext = seal.ext.find('blackjack');
if (!ext) {
    ext = seal.ext.new('blackjack', 'JohNSoN', '1.0.1');
    seal.ext.register(ext);
} else if (ext.version !== VERSION) {
    ext.version = VERSION;
}

const cmd = seal.ext.newCmdItemInfo();
cmd.name = 'bj'; // 指令名字，可用中文
cmd.help = `BlackJack 二十一点 by JohNSoN
.bj reset/init/洗牌
.bj join/加入
.bj quit/退出
.bj start/开始
.bj hit/叫牌
.bj stand/停牌
.bj status/查看/牌桌    查看当前牌桌状态
.bj conclude    (所有玩家结束后)庄家结算. 结算后会自动洗牌.
`;
cmd.disabledInPrivate = true
cmd.solve = (ctx, msg, cmdArgs) => {
    let key = `blackjack:${ctx.group.groupId}`;
    let plID = ctx.player.userId;
    let plName = ctx.player.name;
    let game = new BlackjackGame(ext.storageGet(key));

    let hint = `<${plName}>`;

    let op = cmdArgs.getArgN(1);
    switch (op) {
        case 'reset': case 'init': case '洗牌': {
            game = new BlackjackGame();
            hint += '重置游戏';
            break;
        }
        case 'join': case '加入': {
            let [succ, errMsg] = game.addPlayer(plID, plName);
            if (succ) {
                hint += '加入游戏';
            } else {
                hint += '加入失败: ' + errMsg;
            }
            break;
        }
        case 'quit': case '退出': {
            let [succ, errMsg] = game.removePlayer(plID);
            if (succ) {
                hint += '退出游戏';
            } else {
                hint += '退出失败: ' + errMsg;
            }
            break;
        }
        case 'start': case '开始': {
            let [succ, errMsg] = game.start();
            if (succ) {
                hint += '开始游戏';
                hint += '\n' + game.describe();
            } else {
                hint += '开始失败: ' + errMsg;
            }
            break;
        }
        case 'hit': case '叫牌': {
            let [succ, errMsg] = game.hit(plID);
            if (succ) {
                hint += '叫牌';
            } else {
                hint += '叫牌失败: ' + errMsg;
            }
            hint += '\n' + game.describe();
            break;
        }
        case 'stand': case '停牌': {
            let [succ, errMsg] = game.stand(plID);
            if (succ) {
                hint += '停牌';
            } else {
                hint += '停牌失败: ' + errMsg;
            }
            hint += '\n' + game.describe();
            break;
        }
        case 'conclude': case '结算': {
            let [succ, errMsg] = game.tryConclude();
            if (succ) {
                hint += '结算游戏';
            } else {
                hint += '结算失败: ' + errMsg;
            }
            hint += '\n' + game.describe();

            if (succ) {
                game = new BlackjackGame();
            }
            break;
        }
        case 'status': case '查看': case '牌桌': {
            hint = game.describe();
            break;
        }
        default: {
            const r = seal.ext.newCmdExecuteResult(true);
            r.showHelp = true;
            return r;
        }
    }

    ext.storageSet(key, JSON.stringify(game));
    seal.replyToSender(ctx, msg, hint);
    return seal.ext.newCmdExecuteResult(true);
};
// 将命令注册到扩展中
ext.cmdMap['bj'] = cmd;

function shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
        let j = Math.floor(Math.random() * (i + 1)); // random index from 0 to i
        [array[i], array[j]] = [array[j], array[i]];
    }
}


/*  TODO
    1. 所有人停牌之后自动结算
    2. 结算后显示赢家
    3. (为了仪式感) 玩家自己抽初始手牌
    4. 分牌split
*/


scripts/inventory.js
// ==UserScript==
// @name         物品栏/背包
// @author       JohNSoN
// @version      1.0.0
// @description  记录持有物品的插件
// @timestamp    1679489585
// 2023-03-22 20:53:05
// @license      Unlicense
// @homepageURL  https://github.com/Xiangze-Li/sealdice-addon
// ==/UserScript==

let ext = seal.ext.find('inventory');
if (!ext) {
    ext = seal.ext.new('inventory', 'JohNSoN', '1.0.0');
    seal.ext.register(ext);
}

const cmd = seal.ext.newCmdItemInfo();
cmd.name = 'inventory'; // 指令名字，可用中文
cmd.help = `inventory/i/物品 跑团物品栏
=== 功能 ===
为每个群聊中的每个玩家记录物品栏, 每个条目都可搭配描述信息.
=== 用法 ===
.i show/list/ls/列表
    列出当前物品栏
.i find/cat/查找 <物品名>
    查找名为 <物品名> 的物品
.i add/添加 <物品名> [描述信息]
    添加名为 <物品名> 的物品, 可选地添加描述信息
.i remove/rm/del/移除 <物品名>
    移除名为 <物品名> 的物品
.i modify/mod/edit/修改 <物品名> [描述信息]
    修改名为 <物品名> 的物品的描述信息, 描述信息可以留空
.i clear/clr/清空
    清空物品栏
`;
cmd.solve = (ctx, msg, cmdArgs) => {
    let op = cmdArgs.getArgN(1);
    let key = `inventory:${ctx.group.groupId}:${ctx.player.userId}`;
    let inv = new Map(JSON.parse(ext.storageGet(key) || '[]'));
    let rpl = `<${ctx.player.name}>`;
    let needRefresh = false;
    let retHelp = (err) => {
        if (err) {
            seal.replyToSender(ctx, msg, err);
        }
        const ret = seal.ext.newCmdExecuteResult(true);
        ret.showHelp = true;
        return ret;
    };
    let obj2Str = (obj, desc) => { return `${obj}${desc ? `(${desc})` : ''}` };

    let obj = cmdArgs.getArgN(2);
    let desc = cmdArgs.getRestArgsFrom(3);
    let objEmpty = (obj === undefined || obj === null || obj === '');
    let has = !objEmpty && inv.has(obj);
    let descStorage = has ? inv.get(obj) : '';

    switch (op) {
        case 'show': case 'list': case 'ls': case '列表': {
            if (inv.size > 0) {
                rpl += '目前持有的物品有:\n';
                for (const [k, v] of inv) {
                    rpl += obj2Str(k, v) + '\n';
                }
            } else {
                rpl += '目前没有持有物';
            }
            break;
        }
        case 'find': case 'cat': case '查找': {
            if (objEmpty) return retHelp('物品名为空');
            if (has) {
                rpl += `拥有物品 ${obj2Str(obj, descStorage)}`;
            } else {
                rpl += `没有物品 ${obj}`;
            }
            break;
        }
        case 'add': case '添加': {
            if (objEmpty) return retHelp('物品名为空');
            if (has) {
                rpl += `已经拥有物品 ${obj2Str(obj, descStorage)}`;
            } else {
                inv.set(obj, desc);
                needRefresh = true;
                rpl += `新增物品 ${obj2Str(obj, desc)}`;
            }
            break;
        }
        case 'remove': case 'rm': case 'del': case '移除': {
            if (objEmpty) return retHelp('物品名为空');
            if (has) {
                rpl += `已经删除物品 ${obj2Str(obj, descStorage)}`;
                inv.delete(obj);
                needRefresh = true;
            } else {
                rpl += `没有物品 ${obj}`;
            }
            break;
        }
        case 'modify': case 'mod': case 'edit': case '修改': {
            if (objEmpty) return retHelp('物品名为空');
            if (has) {
                rpl += `修改物品信息 ${obj2Str(obj, descStorage)} => ${desc}`;
                inv.set(obj, desc);
                needRefresh = true;
            } else {
                rpl += `没有物品 ${obj}`;
            }
            break;
        }
        case 'clear': case 'clr': case '清空': {
            inv = new Map();
            needRefresh = true;
            rpl += '已清空物品栏';
            break;
        }
        case 'help': default: {
            return retHelp();
        }
    }

    if (needRefresh) ext.storageSet(key, JSON.stringify([...inv]));

    seal.replyToSender(ctx, msg, rpl);
    return seal.ext.newCmdExecuteResult(true);
};
// 将命令注册到扩展中
ext.cmdMap['inventory'] = cmd;
ext.cmdMap['i'] = cmd;
ext.cmdMap['物品'] = cmd;


scripts/timeline.js
// ==UserScript==
// @name         timeline
// @author       JohNSoN
// @version      1.0.0
// @description  故事时间线跟踪插件
// @timestamp    1678144385
// @license      Unlicense
// @homepageURL  https://github.com/Xiangze-Li/sealdice-addon
// ==/UserScript==

// 首先检查是否已经存在
let ext = seal.ext.find('timeline');
if (!ext) {
    // 不存在，那么建立扩展，名为timeline，作者“JohNSoN”，版本1.0.0
    ext = seal.ext.new('timeline', 'JohNSoN', '1.0.0');
    // 注册扩展
    seal.ext.register(ext);
}

const unitValue = {
    "s": 1000,
    "m": 60 * 1000,
    "h": 60 * 60 * 1000,
    "d": 24 * 60 * 60 * 1000,
};

// 解析时间字符串
//
// 支持格式 `2006-01-02 15:04:05`，时间部分可从后向前省略。将被认为是UTC的
function parseDatetimeString(dateString) {
    let d = new Date(0);

    let sp1 = dateString.split(" ");
    if (sp1.length < 1 || sp1.length > 2) return null;

    let dateSp = sp1[0].split("-");
    if (dateSp.length != 3) return null;
    d.setUTCFullYear(dateSp[0], dateSp[1] - 1, dateSp[2]);

    if (sp1.length == 2) {
        let timeSp = sp1[1].split(":");
        d.setUTCHours(...timeSp)
    } else {
        d.setUTCHours(0, 0, 0, 0);
    }

    return d;
}

// 解析时间间隔字符串
//
// 支持格式 `3D4h5m6s`，任意部分可省略、顺序无关、大小写无关
//
// 返回毫秒数
function parseDurationString(durationString) {
    const regex = /(\d+)([dhms])/g;

    let dur = 0;

    durationString = durationString.toLowerCase();
    let resArr = [...durationString.matchAll(regex)];

    for (let res of resArr) {
        let num = parseInt(res[1]);
        let unit = res[2];
        dur += num * (unitValue[unit] || 0);
    }

    return dur;
}

function durationToString(durationMs) {
    let day = Math.floor(durationMs / unitValue.d);
    durationMs %= unitValue.d;
    let hour = Math.floor(durationMs / unitValue.h);
    durationMs %= unitValue.h;
    let min = Math.floor(durationMs / unitValue.m);
    durationMs %= unitValue.m;
    let sec = Math.floor(durationMs / unitValue.s);

    let reply = "";
    if (day != 0) reply += day + "天";
    if (hour != 0) reply += hour + "时";
    if (min != 0) reply += min + "分";
    if (sec != 0) reply += sec + "秒";

    return (reply !== "" ? reply : "0");
}

function toReplyString(storage) {
    if (storage.baseSet) {
        return "当前时间为 " + formatDate(new Date(storage.base + storage.delta));
    }
    return "当前经过时间为 " + durationToString(storage.delta) + "(未设置基准时间)";
}

function formatDate(dateObj) {
    let year = dateObj.getUTCFullYear();
    let month = dateObj.getUTCMonth() + 1;
    let date = dateObj.getUTCDate();

    let day = dateObj.getUTCDay();
    const dayMap = ["日", "一", "二", "三", "四", "五", "六"];
    day = "星期" + dayMap[day];

    let hour = dateObj.getUTCHours();
    if (hour < 10) hour = "0" + hour;
    let min = dateObj.getUTCMinutes();
    if (min < 10) min = "0" + min;
    let sec = dateObj.getUTCSeconds();
    if (sec < 10) sec = "0" + sec;

    return `${year}年${month}月${date}日(${day}) ${hour}:${min}:${sec}`;
}

const cmd = seal.ext.newCmdItemInfo();
cmd.name = 'timeline'; // 指令名字，可用中文
cmd.help =
    `timeline 故事时间线跟踪工具
==========================
.timeline setbase <now|基准时间>
    设置基准时间，时间格式为 2006-01-02 15:04:05，时间部分可从后向前省略
.timeline <add|sub> <时间间隔>
    增加或减少经过时间，时间间隔格式为 2D3h4m5s，任意部分可省略
.timeline forward <时间点>
    将当前时间向前推进到指定时间点，需先设置基准时间，不能回溯
.timeline reset
    重置经过时间为0
.timeline <rm|del|clr>
    清除基准时间和经过时间
.timeline show
    显示当前时间线
.timeline help
    显示帮助信息`;

cmd.solve = (ctx, msg, cmdArgs) => {
    const key = "timeline:" + ctx.group.groupId;
    let op = cmdArgs.getArgN(1);

    let st = JSON.parse(ext.storageGet(key) || "{}");
    if (st.base === undefined) st.base = 0;
    if (st.baseSet === undefined) st.baseSet = false;
    if (st.delta === undefined) st.delta = 0;

    let reply = "<" + ctx.player.name + ">";

    switch (op) {
        case "setbase": {
            let date = cmdArgs.getRestArgsFrom(2);
            let dateObj = new Date(Date.now() - (new Date()).getTimezoneOffset() * 60 * 1000);
            if (date != "now") {
                dateObj = parseDatetimeString(date);
                if (dateObj === null) {
                    reply = "无法解析时间字符串 " + date + "\n";
                    break;
                }
            }
            st.base = dateObj.getTime();
            st.baseSet = true;
            reply += "设置基准时间为 " + formatDate(dateObj) + "\n";
            reply += toReplyString(st) + "\n";
            break;
        }
        case "add": case "sub": {
            let dur = cmdArgs.getRestArgsFrom(2);
            let durMs = parseDurationString(dur);
            if (op == 'sub' && durMs > st.delta) {
                reply += "尝试减少经过时间" + durationToString(durMs) + "失败\n";
                reply += "变化量超过了当前经过时间 (" + durationToString(st.delta) + ")\n";
                break;
            }
            st.delta += (op == "add" ? 1 : -1) * durMs;
            reply += (op == "add" ? "增加" : "减少") + "经过时间" + durationToString(durMs) + "\n";
            reply += toReplyString(st) + "\n";
            break;
        }
        case "forward": case "fwd": {
            let date = cmdArgs.getRestArgsFrom(2);
            let dateObj = parseDatetimeString(date);
            if (dateObj === null) {
                reply = "无法解析时间字符串 " + date + "\n";
                break;
            }
            if (!st.baseSet) {
                reply += "尝试将时间线前进到 " + formatDate(dateObj) + " 失败\n";
                reply += "未设置基准时间，无法进行时间线前进\n";
                break;
            }
            let dateTs = dateObj.getTime();
            if (dateTs <= st.base + st.delta) {
                reply += "尝试将时间线前进到 " + formatDate(dateObj) + " 失败\n";
                reply += "目标时间必须晚于当前时间 (" + +")\n";
                break;
            }
            st.delta = dateTs - st.base;
            reply += "将时间线前进到 " + formatDate(dateObj) + "\n";
            break;
        }
        case "reset": {
            st.delta = 0;
            reply += "已重置经过时间\n" + toReplyString(st) + "\n";
            break;
        }
        case "rm": case "del": case "clr": {
            st.base = 0;
            st.baseSet = false;
            st.delta = 0;
            reply += "已清空时间线\n";
            break;
        }
        case "show": {
            reply = toReplyString(st) + "\n";
            break;
        }
        case 'help': default: {
            const ret = seal.ext.newCmdExecuteResult(true);
            ret.showHelp = true;
            return ret;
        }
    }

    ext.storageSet(key, JSON.stringify(st));
    seal.replyToSender(ctx, msg, reply);
    return seal.ext.newCmdExecuteResult(true);
};
// 将命令注册到扩展中
ext.cmdMap['timeline'] = cmd;
